sys/net/application_layer/nanocoap/sock.c

```diff
@@ -41,59 +55,90 @@ int nanocoap_sock_connect(nanocoap_sock_t *sock, sock_udp_ep_t *local, sock_udp_
     return sock_udp_create(sock, local, remote, 0);
 }
 
-ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len)
+static int _get_error(const coap_pkt_t *pkt)
+{
+    switch (coap_get_code_class(pkt)) {
+    case COAP_CLASS_CLIENT_FAILURE:
+        return -ENXIO;
+    case COAP_CLASS_SERVER_FAILURE:
+        return -EREMOTE;
+    default:
+        return 0;
+    }
+}
+
+static int _send_ack(nanocoap_sock_t *sock, coap_pkt_t *pkt)
+{
+    coap_hdr_t ack;
+    unsigned tkl = coap_get_token_len(pkt);
+
+    coap_build_hdr(&ack, COAP_TYPE_ACK, pkt->token, tkl,
+                   COAP_CODE_VALID, ntohs(pkt->hdr->id));
+
+    return sock_udp_send(sock, &ack, sizeof(ack), NULL);
+}
+
+ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
+                                 coap_request_cb_t cb, void *arg)
 {
-    ssize_t res = -EAGAIN;
+    ssize_t tmp, res = 0;
     size_t pdu_len = (pkt->payload - (uint8_t *)pkt->hdr) + pkt->payload_len;
     uint8_t *buf = (uint8_t*)pkt->hdr;
     uint32_t id = coap_get_id(pkt);
+    void *payload, *ctx = NULL;
+
+    unsigned state = STATE_SEND_REQUEST;
 
-    /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT *
-     * ACK_RANDOM_FACTOR) */
+    /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT * ACK_RANDOM_FACTOR) */
```

Comment:
This one really should be addressed, but as it is pre-existing, I guess it deserves a dedicated PR.